Applications/Spire/Source/OrderImbalanceIndicator/CachedOrderImbalanceIndicatorModel.cpp

```diff
@@ -0,0 +1,116 @@
+#include "Spire/OrderImbalanceIndicator/CachedOrderImbalanceIndicatorModel.hpp"
+
+using namespace boost;
+using namespace boost::icl;
+using namespace boost::posix_time;
+using namespace Nexus;
+using namespace Spire;
+
+CachedOrderImbalanceIndicatorModel::CachedOrderImbalanceIndicatorModel(
+    std::shared_ptr<OrderImbalanceIndicatorModel> source)
+    : m_source_model(std::move(source)) {
+  auto [connection, promise] = m_source_model->subscribe(
+    [=] (const auto& imbalance) {
+      on_imbalance_published(imbalance);
+    });
+  m_subscription_connection = std::move(connection);
+}
+
+QtPromise<std::vector<Nexus::OrderImbalance>>
+    CachedOrderImbalanceIndicatorModel::load(
+    const TimeInterval& interval) {
+  if(boost::icl::contains(m_ranges, interval)) {
+    return load_from_cache(interval);
+  }
+  return load_from_model(interval);
+}
+
+SubscriptionResult<optional<Nexus::OrderImbalance>>
+    CachedOrderImbalanceIndicatorModel::subscribe(
+    const OrderImbalanceSignal::slot_type& slot) {
+  return m_source_model->subscribe(slot);
+}
+
+std::shared_ptr<OrderImbalanceChartModel>
+    CachedOrderImbalanceIndicatorModel::get_chart_model(
+    const Security& security) {
+  throw std::runtime_error("method not implemented");
+}
+
+QtPromise<std::vector<Nexus::OrderImbalance>>
+    CachedOrderImbalanceIndicatorModel::load_from_cache(
+    const TimeInterval& interval) {
+  auto first = [&] {
```

Comment:
Use a local model for your implementation. Part of the benefit of having a local/in-memory model is that you can use it to create more sophisticated models.

Reply:
Should I add an insert method to the local model that takes a vector? Or just insert the new imbalances one-by-one?

Reply:
One by one is fine.